### DISCUSSION TITLE
avoid objects from penetrating

### DISCUSSION THREAD

**Discussion by shivanimall** [bulletphysics/bullet3#3858] [Q&A] [answered]
Hello @erwincoumans,

I have 2 objects meshes, A and B- such that I place object B at P_b = (0, 0, 0) and O_b= (0, 0, 0, 1). and place object A at position P and orientation O as follows:
P = vertex coordinate(VC) from object mesh B and O = Euler angle determined by vertex normal direction (VN) of vertex coordinate (VC) also from object mesh B.

Few things I notice are:
1/ sometimes with the preset P and O object A has already penetrated into the object B by some percentage.
2/ at this preset P and O when I apply a force while stepping through simulation, in many cases I see object A further penetrate through object B.

In my URDF file I make sure to define both collision and visual geometry for both the object. However I use the same mesh file, which I suspect may be the issue.

Any hints on where I can look into for debugging this to avoid any such penetration of one object into another? or what maybe causing this?
Here are images attached for your convenience:
![image](https://user-images.githubusercontent.com/5581525/118968432-8c462f00-b989-11eb-8929-b86ee74a17ed.png)
![image](https://user-images.githubusercontent.com/5581525/118968693-d7604200-b989-11eb-9089-b762946a0dd1.png)

Please let me know if my question is unclear or if you need more details. Thank you!

**Comment by erwincoumans** [comment, marked as answer]
Use box collision shapes (or convex shapes) if possible, and use a small time step (default is 1./240 or 240 Hz, you could go smaller, such as 1./1000 or 1000Hz).

**Reply by shivanimall** [reply]
thanks Erwin, let me try out and update.
